# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [3]:
def add_two(x):
    return x + 2 
print(add_two(3))
print(add_two)

5
<function add_two at 0x7f4f2ca54ee0>


In [4]:
lambda x: x + 2

(lambda x: x + 2)(3)

5

#### Saving to a Variable

In [5]:
f = lambda x, y:x+y+2
f(3,2)

7

#### Multiple Inputs

In [7]:
l = 3
g = 5
p = 8
(lambda x, y, z: x*y*z)(l,g,p)


120

#### Passing a Lambda into a Function

In [8]:
def do_math(func, num):
    return func(num)

do_math(lambda x: x*x, 4)

16

In [9]:
def do_math(func, num):
    return func(num)

def my_math(x):
    return x*x
do_math(my_math, 4)

16

#### Returning a Lambda from a Function

In [15]:
def return_func():
    def squared(num):
        return num**2
    return squared

print(type(return_func()))
print(return_func()(3))

def return_func():
    return lambda num1, num2: num1**2+num2**2
print(type(return_func()))
print(return_func()(3,6))

<class 'function'>
9
<class 'function'>
45


#### If Statements within Lambdas

In [17]:
#Lambda x : True if (condition) else false
f = lambda num: num*2 if num > 10 else num + 2

f(8)

10

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [18]:
f = lambda x: x**3
f(4)

64

In [19]:
#get good at reading Documentation because this is your lifeline also
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [23]:
#Using Lambdas with sort to sort a list of dictionary by a key from highest to smallest
from random import randint

my_class = [
    {"name":"Jody", "IQ":randint(1,240)},
    {"name":"Sarah", "IQ":randint(1,240)},
    {"name":"Dylan", "IQ":randint(1,240)},
    {"name":"Steven", "IQ":randint(1,240)}
]

print(my_class)
print(sorted(my_class, key=lambda student: student["IQ"], reverse=True ))

[{'name': 'Jody', 'IQ': 197}, {'name': 'Sarah', 'IQ': 139}, {'name': 'Dylan', 'IQ': 190}, {'name': 'Steven', 'IQ': 36}]
[{'name': 'Jody', 'IQ': 197}, {'name': 'Dylan', 'IQ': 190}, {'name': 'Sarah', 'IQ': 139}, {'name': 'Steven', 'IQ': 36}]


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [24]:
# map(func,iterable(list,dict,tuple,set,etc...))
# Normally the usage of map happens with a pre-defined function -- but can use lambdas as well
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [28]:
def squared(num, num2):
    if num < 10 and num2 <10:
        return (num**2, num2**2)
    else:
        return (num, num2)
numbers = [4,10,3,2,6]
more_nums = [4,1,20,3,15]

new_list=[]
for i in range(len(numbers)):
    new_list.append(squared(numbers[i], more_nums[i]))
new_list




[(16, 16), (10, 1), (3, 20), (4, 9), (6, 15)]

In [30]:
squared_nums_map = map(squared,numbers,more_nums)
list(squared_nums_map)

[(16, 16), (10, 1), (3, 20), (4, 9), (6, 15)]

#### Using Lambda's with Map

In [31]:
# map(Lambda x+2,list)
# using Lambda in map happens in line (or in one line)
# list(map(lambda x: x*2,nums))
numbers = [4,10,3,2,6]
squared_nums=list(map(lambda x: x**2 if x<10 else x, numbers))
print(squared_nums)

[16, 10, 9, 4, 36]


In [34]:
numbers = [4,10,3,2,6]
more_nums = [4,1,20,3,15]
squared_nums2 = list(map((lambda x,y: (x**2, y**2) if x<10 and y<10 else (x, y)), numbers, more_nums))
print(squared_nums2)

[(16, 16), (10, 1), (3, 20), (4, 9), (6, 15)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [39]:
nums = [4,2,6,8]

dub_minus_one = list(map(lambda num: (2*num)-1, nums))
print(dub_minus_one)
# output
# [7,3,11,15]
print(nums)

[7, 3, 11, 15]
[4, 2, 6, 8]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [35]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [38]:
bool('')

False

In [36]:
list(filter(None, [True,True, False, False, "", "string"]))

[True, True, 'string']

#### Using Lambda's with Filter()

In [40]:
names = ["Shayne", "Gio", "Sarah", "Steven", "Francisco", "Dmitry"]

def s_names(name):
    if name[0].lower()=='s':
        return True
    else:
        return False

new_names_list = list(filter(s_names, names))
print(new_names_list)


['Shayne', 'Sarah', 'Steven']


In [42]:
new_names_lamb = list(filter(lambda name: True if name[0].lower()=='s' else False, names))
new_names_lamb

['Shayne', 'Sarah', 'Steven']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [44]:
from statistics import mean

nums =[2,7,4.2,1.6,9,4.4,4.9]
y=mean(nums)

below_mean = list(filter(lambda x: True if x <= y else False ,nums))
print(mean(nums))
print(below_mean)


4.728571428571429
[2, 4.2, 1.6, 4.4]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [46]:
from functools import reduce
# reduce(function,iterable)
help(reduce)


Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, iterable[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence
    or iterable, from left to right, so as to reduce the iterable to a single
    value.  For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the iterable in the calculation, and serves as a default when the
    iterable is empty.



In [48]:
list_1 = [2, 4, 6, 8, 10]
def subtract_nums(num1, num2):
    return num1 - num2

result = reduce(subtract_nums, list_1)
print(result)

-26


#### Using Lambda's with Reduce()

In [53]:
list_2 = [1,10,2,30]
result_lamb = reduce(lambda num1, num2: num1-num2, list_2)
result_lamb

-41

In [54]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [56]:
l_5  = [1,2,3,4]
result_5 = reduce((lambda aggregator, current_value: aggregator * current_value), l_5)

print(result_5)

24


In [60]:
print("*"*5*3)

***************


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [64]:
def recursive(i):
    if i == 4: # Base case
        return
    else:
        print("hi! ")
        i+=1
        recursive(i)
        

recursive(0)

hi! 
hi! 
hi! 
hi! 


#### Implementing a Base Case

In [66]:
def add_nums(num):
    if num <= 1:
        print("add_nums(1)=1")
        return num
    else:
        print(f"add_nums({num}) = {num} + add_nums({num-1})")
        x = num - 1
        return num + add_nums(x)
add_nums(5)

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1)=1


15

#### Writing a Factorial Function

In [70]:
# 5! = 5*4*3*2*1
def factorial(num):
    if num <= 1:
        print('factorial(1)=1')
        return 1
    else:
        print(f"factorial({num})={num} * factorial({num-1})")
        return num * factorial(num-1)
factorial(10)


factorial(10)=10 * factorial(9)
factorial(9)=9 * factorial(8)
factorial(8)=8 * factorial(7)
factorial(7)=7 * factorial(6)
factorial(6)=6 * factorial(5)
factorial(5)=5 * factorial(4)
factorial(4)=4 * factorial(3)
factorial(3)=3 * factorial(2)
factorial(2)=2 * factorial(1)
factorial(1)=1


3628800

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [71]:
def sub_nums(num):
    if num <= 1:
        print("add_nums(1)=1")
        return num
    else:
        print(f"add_nums({num}) = {num} - add_nums({num-1})")
        x = num - 1
        return num - sub_nums(x)
sub_nums(5)



add_nums(5) = 5 - add_nums(4)
add_nums(4) = 4 - add_nums(3)
add_nums(3) = 3 - add_nums(2)
add_nums(2) = 2 - add_nums(1)
add_nums(1)=1


3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

In [73]:
list(range(5))

[0, 1, 2, 3, 4]

In [75]:
def test():
    yield 1
    yield 2
    yield 3
    yield "a"
    yield ["a","b","c"]

for i in test():
    print(i)
    
print(list(test()))

1
2
3
a
['a', 'b', 'c']
[1, 2, 3, 'a', ['a', 'b', 'c']]


#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [76]:
def my_range(n):
    for i in range(n):
        yield i
list(my_range(5))


[0, 1, 2, 3, 4]

In [77]:
def my_range(start, stop, step=1):
    while start<stop:
        yield start
        start+=step

list(my_range(1,5,1))
        

[1, 2, 3, 4]

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

In [81]:
def range_squared(start, stop, step=1):
    while start > stop:
        yield start ** 2
        start -= step
    

for i in range_squared(10, 0, 1):
    print(i)
# 100
# 81
# 64
# 49
# 36
# 25
# 16
# 9
# 4
# 1


100
81
64
49
36
25
16
9
4
1


#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [82]:
help(list.sort)

Help on method_descriptor:

sort(self, /, *, key=None, reverse=False)
    Sort the list in ascending order and return None.
    
    The sort is in-place (i.e. the list itself is modified) and stable (i.e. the
    order of two equal elements is maintained).
    
    If a key function is given, apply it once to each list item and sort them,
    ascending or descending, according to their function values.
    
    The reverse flag can be set to sort in descending order.



# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [2]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

places_filter = list(filter((lambda city: city if len(city) > 0 and city[0].isalpha() else False), places))

print(places_filter)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [1]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key = (lambda name: name.split()[-1].lower()))

print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [56]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

places_f = list(map((lambda city: (city[0], 9 / 5 * city[1] + 32)), places))

print(places_f)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [4]:
def fib(n):
   
    if n == 1:
        return 1
    elif n == 2:
        return 1
    elif n > 2:
        return fib(n - 1) + fib(n - 2)
    
for n in range(1, 7):
    print(f"Iteration {n - 1}: {fib(n)}")

Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
